### PR TITLE
Initialize iframe viewport immediately

### LINKED ISSUE #14364: Transitory initial width and height for windows in iframes
When an iframe is first created, its `window_size` is set to `None`. This is later set by `Constellation::handle_iframe_size_message` (https://github.com/servo/servo/blob/master/components/constellation/constellation.rs#L1312) in response to a `FrameSize` event from the compositor, which contains the size calculated by webrender in `RenderNotifier::pipeline_size_changed` (https://github.com/servo/servo/blob/master/components/compositing/compositor.rs#L334).

This is fine, except that it means script may be able to see a transitory zero value for the calculated height and width, if it queries the window size before webrender has reported.

Reported by @Ms2ger on irc: http://logs.glob.uno/?c=mozilla%23servo&s=24+Nov+2016&e=24+Nov+2016#c568181

### LINKED ISSUE #15688: Iframe's window size is not available synchronously in about:blank
```html
<iframe></iframe>
<script>console.log(document.querySelector('iframe').contentWindow.innerWidth)</script>
```
This prints 0, because the `NewLayoutInfo` in `HTMLIFrameElement::navigate_or_reload_child_browsing_context` provides a `None` value for `window_size`, which is what is stored in the new `Window` that is created and retrieved by `Window::GetInnerWidth`.

### LINKED ISSUE #15689: Transparent Rects demo is confined to top left corner
If you run `./mach run --release -b` and click the transparent rects demo, all of the rects congregate in the top left corner. If you run `./mach run --release https://cdn.rawgit.com/servo/servo.org/gh-pages/bhtml-newtab/webrender-demos/transparent_rects.html` it works fine. 

The constellation creates the new pipeline in `handle_script_loaded_url_in_iframe_msg` with a null window size, which gets filled in later when the iframe element is reflowed and sends a message to the constellation. If the iframe's content loads before this occurs (such as when using file:// URLs), the script will run before a resize message is received, so `Window::GetInnerWidth` will return 0 (similar to, but not exactly #15688).

### PR DESCRIPTION
**Pull request by jdm**
This is probably a significant cause of unstable test results, and it finally bothered me enough to fix it. The solution isn't great for performance (a sync layout query every time a iframe is added to a document), but performance needs to follow correctness.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #14364 and fix #15689 and fix #15688.
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22395)

### CODE REVIEW

**Comment by jdm** [comment]
Who could have foreseen that invoking layout inside of a method that can be called while the document is in an unstable state (ie. from the Node::children_changed callback, which is invoked during the Node::insert algorithm) could lead to layout's head exploding?

**Review by Manishearth — APPROVED** [review]
r=me

**Comment by jdm** [comment]
@bors-servo retry
* #22463

**Comment by jdm** [comment]
@bors-servo r-
```
  ▶ CRASH [expected OK] /html/syntax/parsing/html5lib_entities01.html?run_type=write
  │
  │ assertion failed: `(left != right)`
  │   left: `0`,
  │  right: `0` (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(2), index: PipelineIndex(2) }, at components/gfx/platform/macos/font.rs:250)
  │ stack backtrace:
  │    0:        0x10bc50504 - __ZN9backtrace9backtrace5trace17hefbd7c3b17ad6e6eE
  │                         at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/backtrace/mod.rs:42
  │    1:        0x10bc495ac - __ZN9backtrace7capture9Backtrace14new_unresolved17hfa7d39a5ef5fb84fE
  │                         at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:88
  │    2:        0x10bc4950e - __ZN9backtrace7capture9Backtrace3new17hb34b0ddcba79179eE
  │                         at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:63
  │    3:        0x105d216b0 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17ha0f5a2fdbedb9aa5E
  │                         at ports/servo/non_android_main.rs:111
  │    4:        0x10da13468 - __ZN3std9panicking20rust_panic_with_hook17h14490a1041ee4235E
  │                         at libstd/panicking.rs:480
  │    5:        0x10da12f7c - __ZN3std9panicking18continue_panic_fmt17h7beddad0ba45f4eaE
  │                         at libstd/panicking.rs:390
  │    6:        0x10da12ed0 - __ZN3std9panicking15begin_panic_fmt17h19527e965f98f107E
  │                         at libstd/panicking.rs:345
  │    7:        0x109ea0b97 - __ZN87_$LT$gfx..platform..macos..font..FontHandle$u20$as$u20$gfx..font..FontHandleMethods$GT$11glyph_index17hab28a653bac06806E
  │                         at components/gfx/platform/macos/font.rs:250
  │    8:        0x109ea31bf - __ZN3gfx4font4Font11glyph_index17h723bf512f541bff6E
  │                         at components/gfx/font.rs:327
  │    9:        0x109ea320a - __ZN3gfx4font4Font13has_glyph_for17h7315b167bece5cf6E
  │                         at components/gfx/font.rs:331
  │   10:        0x1092ccf9f - __ZN3gfx4font9FontGroup17find_by_codepoint28_$u7b$$u7b$closure$u7d$$u7d$17h6d77c181c5c874a2E
  │                         at /Users/jdm/src/servo/components/gfx/font.rs:391
  │   11:        0x1092cd007 - __ZN3gfx4font9FontGroup17find_by_codepoint28_$u7b$$u7b$closure$u7d$$u7d$17h96780a700a5fa211E
  │                         at /Users/jdm/src/servo/components/gfx/font.rs:393
  │   12:        0x1092a944c - __ZN4core4iter8iterator8Iterator4find28_$u7b$$u7b$closure$u7d$$u7d$17h69032f744bf9c722E
  │                         at libcore/iter/iterator.rs:1783
  │   13:        0x1092a8cb4 - __ZN4core4iter8iterator8Iterator12try_for_each28_$u7b$$u7b$closure$u7d$$u7d$17h4b85a7d0f73c0399E
  │                         at libcore/iter/iterator.rs:1559
  │   14:        0x109350896 - __ZN90_$LT$core..iter..FilterMap$LT$I$C$$u20$F$GT$$u20$as$u20$core..iter..iterator..Iterator$GT$8try_fold28_$u7b$$u7b$closure$u7d$$u7d$17h18e65ac81d83c910E
  │                         at libcore/iter/mod.rs:1659
  │   15:        0x1093ba4a3 - __ZN94_$LT$core..slice..IterMut$LT$$u27$a$C$$u20$T$GT$$u20$as$u20$core..iter..iterator..Iterator$GT$8try_fold17h2cdbe62382c6914aE
  │                         at libcore/slice/mod.rs:2720
  │   16:        0x109350797 - __ZN90_$LT$core..iter..FilterMap$LT$I$C$$u20$F$GT$$u20$as$u20$core..iter..iterator..Iterator$GT$8try_fold17he3cbd0efa9c17ed2E
  │                         at libcore/iter/mod.rs:1658
  │   17:        0x109342fb4 - __ZN4core4iter8iterator8Iterator12try_for_each17ha3e84e8eed578b50E
  │                         at libcore/iter/iterator.rs:1559
  │   18:        0x1093433e4 - __ZN4core4iter8iterator8Iterator4find17hffb2c1cffd3acfadE
  │                         at libcore/iter/iterator.rs:1782
  │   19:        0x1092cd15b - __ZN3gfx4font9FontGroup4find17haa5aa0b5a7260c7fE
  │                         at /Users/jdm/src/servo/components/gfx/font.rs:429
  │   20:        0x1092ccd1e - __ZN3gfx4font9FontGroup17find_by_codepoint17h6c74e2ff1df5c97fE
  │                         at /Users/jdm/src/servo/components/gfx/font.rs:393
  │   21:        0x1093543ee - __ZN6layout4text14TextRunScanner19flush_clump_to_list17h6906fbc670fd5860E
  │                         at components/layout/text.rs:236
  │   22:        0x1093530d5 - __ZN6layout4text14TextRunScanner13scan_for_runs17ha8b6c6b52237b044E
  │                         at components/layout/text.rs:126
  │   23:        0x1064beb7a - __ZN93_$LT$layout..construct..FlowConstructor$LT$$u27$a$C$$u20$ConcreteThreadSafeLayoutNode$GT$$GT$30flush_inline_fragments_to_flow28_$u7b$$u7b$closure$u7d$$u7d$17hfe8263542e2a87aeE
  │                         at /Users/jdm/src/servo/components/layout/construct.rs:509
  │   24:        0x106678840 - __ZN6layout7context30with_thread_local_font_context28_$u7b$$u7b$closure$u7d$$u7d$17h7b9953fc7bc2553cE
  │                         at /Users/jdm/src/servo/components/layout/context.rs:44
  │   25:        0x10652f93f - __ZN46_$LT$std..thread..local..LocalKey$LT$T$GT$$GT$8try_with17h8fd818a0e5c73765E
  │                         at libstd/thread/local.rs:294
  │   26:        0x10652caf6 - __ZN46_$LT$std..thread..local..LocalKey$LT$T$GT$$GT$4with17h20665618f1332d0aE
  │                         at libstd/thread/local.rs:248
  │   27:        0x106677cb5 - __ZN6layout7context30with_thread_local_font_context17h0572c38ab81d9668E
  │                         at /Users/jdm/src/servo/components/layout/context.rs:38
  │   28:        0x1064c080c - __ZN93_$LT$layout..construct..FlowConstructor$LT$$u27$a$C$$u20$ConcreteThreadSafeLayoutNode$GT$$GT$44build_flow_for_block_starting_with_fragments17h12ea0d9c8b432b78E
  │                         at /Users/jdm/src/servo/components/layout/construct.rs:718
  │   29:        0x1064bd3c7 - __ZN93_$LT$layout..construct..FlowConstructor$LT$$u27$a$C$$u20$ConcreteThreadSafeLayoutNode$GT$$GT$25build_flow_for_block_like17h50deea658c63213fE
  │                         at /Users/jdm/src/servo/components/layout/construct.rs:803
  │   30:        0x1064ba646 - __ZN93_$LT$layout..construct..FlowConstructor$LT$$u27$a$C$$u20$ConcreteThreadSafeLayoutNode$GT$$GT$20build_flow_for_block17h6f4734f6b5c60a9bE
  │                         at /Users/jdm/src/servo/components/layout/construct.rs:881
  │   31:        0x10649676c - __ZN185_$LT$layout..construct..FlowConstructor$LT$$u27$a$C$$u20$ConcreteThreadSafeLayoutNode$GT$$u20$as$u20$layout..traversal..PostorderNodeMutTraversal$LT$ConcreteThreadSafeLayoutNode$GT$$GT$7process17ha4a20e4ad8656025E
  │                         at /Users/jdm/src/servo/components/layout/construct.rs:1942
  │   32:        0x106692b86 - __ZN6layout9traversal18construct_flows_at17hfaa395b8db02d8dbE
  │                         at /Users/jdm/src/servo/components/layout/traversal.rs:210
  │   33:        0x10667ff33 - __ZN121_$LT$layout..traversal..RecalcStyleAndConstructFlows$LT$$u27$a$GT$$u20$as$u20$style..traversal..DomTraversal$LT$E$GT$$GT$17process_postorder17h47f0a10a00a897a3E
  │                         at /Users/jdm/src/servo/components/layout/traversal.rs:74
  │   34:        0x106690292 - __ZN5style9traversal12DomTraversal26handle_postorder_traversal17hc6636c7e2cc41379E
  │                         at /Users/jdm/src/servo/components/style/traversal.rs:119
  │   35:        0x1065b3ccd - __ZN5style6driver12traverse_dom17h9d5bbe9211c283caE
  │                         at /Users/jdm/src/servo/components/style/driver.rs:116
  │   36:        0x10651f939 - __ZN13layout_thread12LayoutThread13handle_reflow28_$u7b$$u7b$closure$u7d$$u7d$17h26006c2fa063068fE
  │                         at components/layout_thread/lib.rs:1398
  │   37:        0x10655c7d3 - __ZN14profile_traits4time7profile17h75e9edba9ed94cc7E
  │                         at /Users/jdm/src/servo/components/profile_traits/time.rs:142
  │   38:        0x10651e3a3 - __ZN13layout_thread12LayoutThread13handle_reflow17h4cc1b33e70c34ec3E
  │                         at components/layout_thread/lib.rs:1392
  │   39:        0x106517e9c - __ZN13layout_thread12LayoutThread21handle_request_helper28_$u7b$$u7b$closure$u7d$$u7d$17h939e502e8b3a706aE
  │                         at components/layout_thread/lib.rs:748
  │   40:        0x10655bdb3 - __ZN14profile_traits4time7profile17h2831e0b217ffc080E
  │                         at /Users/jdm/src/servo/components/profile_traits/time.rs:142
  │   41:        0x10651662b - __ZN13layout_thread12LayoutThread21handle_request_helper17hcb9b165d0c350c01E
  │                         at components/layout_thread/lib.rs:744
  │   42:        0x106515546 - __ZN13layout_thread12LayoutThread14handle_request17hf674096a0bf8f45bE
  │                         at components/layout_thread/lib.rs:694
  │   43:        0x10651433c - __ZN13layout_thread12LayoutThread5start17h839677ce7ab45c91E
  │                         at components/layout_thread/lib.rs:580
  │   44:        0x106510fc4 - __ZN82_$LT$layout_thread..LayoutThread$u20$as$u20$layout_traits..LayoutThreadFactory$GT$6create28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h517d252de8a5338cE
  │                         at components/layout_thread/lib.rs:318
  │   45:        0x10655b365 - __ZN14profile_traits3mem12ProfilerChan25run_with_memory_reporting17h9df8712025d1ba68E
  │                         at /Users/jdm/src/servo/components/profile_traits/mem.rs:88
  │   46:        0x106511681 - __ZN82_$LT$layout_thread..LayoutThread$u20$as$u20$layout_traits..LayoutThreadFactory$GT$6create28_$u7b$$u7b$closure$u7d$$u7d$17h90fe7f56bb92753fE
  │                         at components/layout_thread/lib.rs:316
  │   47:        0x1066f42c4 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h01e2ddcbf2ab7975E
  │                         at libstd/sys_common/backtrace.rs:136
  │   48:        0x1064f18b4 - __ZN3std6thread7Builder5spawn28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h8b70c031fb5ed115E
  │                         at libstd/thread/mod.rs:409
  │   49:        0x1065255e4 - __ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17h665c3b5eb75181fcE
  │                         at libstd/panic.rs:313
  │   50:        0x1066f4f34 - __ZN3std9panicking3try7do_call17h1bca9e2403ab94abE
  │                         at libstd/panicking.rs:310
  │   51:        0x10da1f9de - ___rust_maybe_catch_panic
  │                         at libpanic_unwind/lib.rs:102
  │   52:        0x1066f4c33 - __ZN3std9panicking3try17he4162fb5191d4011E
  │                         at libstd/panicking.rs:289
  │   53:        0x10652c534 - __ZN3std5panic12catch_unwind17h2f968a3a2ce5ec1bE
  │                         at libstd/panic.rs:392
  │   54:        0x1064f16de - __ZN3std6thread7Builder5spawn28_$u7b$$u7b$closure$u7d$$u7d$17h81ac2bc815305067E
  │                         at libstd/thread/mod.rs:408
  │   55:        0x1064f45e3 - __ZN50_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$8call_box17hb83335dcb39ffcaeE
  │                         at liballoc/boxed.rs:672
  │   56:        0x10d9f697b - __ZN3std3sys4unix6thread6Thread3new12thread_start17hba53abd905cebc29E
  │                         at libstd/sys_common/thread.rs:24
  │   57:     0x7fffd4c7e93a - __pthread_body
  │   58:     0x7fffd4c7e886 - __pthread_start
  │ [2018-12-14T20:08:56Z ERROR servo] assertion failed: `(left != right)`
  │   left: `0`,
  │  right: `0`
  └ Pipeline failed in hard-fail mode.  Crashing!
```

**Comment by CYBAI** [comment]
@bors-servo retry

```
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Overall test", 
    "test": "/_webgl/conformance/ogles/GL/refract/refract_001_to_006.html", 
    "line": 51569, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/_webgl/conformance/ogles/GL/refract/refract_001_to_006.html", 
    "line": 51572, 
    "action": "test_result", 
    "expected": "OK"
}
```

**Comment by CYBAI** [comment]
@bors-servo retry

- Another WebGL intermittent

```
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Overall test", 
    "test": "/_webgl/conformance/ogles/GL/build/build_161_to_168.html", 
    "line": 57699, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/_webgl/conformance/ogles/GL/build/build_161_to_168.html", 
    "line": 57706, 
    "action": "test_result", 
    "expected": "OK"
}
```

**Comment by CYBAI** [comment]
@bors-servo retry

```
{
    "status": "ERROR", 
    "group": "default", 
    "message": "The operation is not supported.", 
    "stack": "undefined:undefined:undefined", 
    "subtest": null, 
    "test": "/html/semantics/embedded-content/media-elements/playing-the-media-resource/pause-move-within-document.html", 
    "line": 126945, 
    "action": "test_result", 
    "expected": "OK"
}
```

```
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/html/webappapis/animation-frames/cancel-invoked.html", 
    "line": 126767, 
    "action": "test_result", 
    "expected": "OK"
}
```

**Comment by jdm** [comment]
@bors-servo retry
* network failure

**Comment by CYBAI** [comment]
@bors-servo retry

```
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/html/semantics/tabular-data/the-table-element/table-insertRow.html", 
    "line": 135259, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/resource-timing/resource_TAO_wildcard.htm", 
    "line": 135640, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/navigation-timing/dom_interactive_media_document.html", 
    "line": 136831, 
    "action": "test_result", 
    "expected": "ERROR"
}
```

**Comment by CYBAI** [comment]
@bors-servo retry

```
{
  "status": "TIMEOUT",
  "group": "default",
  "message": null,
  "stack": null,
  "subtest": null,
  "test": "/html/semantics/embedded-content/the-iframe-element/same_origin_parentage.html",
  "line": 135833,
  "action": "test_result",
  "expected": "OK"
}
```

**Comment by CYBAI** [comment]
@bors-servo retry

```
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/wasm/jsapi/table/constructor.any.html", 
    "line": 132064, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/performance-timeline/po-getentries.any.worker.html", 
    "line": 132152, 
    "action": "test_result", 
    "expected": "OK"
}
```

File: components/script/dom/htmlscriptelement.rs [hunk 1]
```diff
@@ -777,7 +777,10 @@ impl VirtualMethods for HTMLScriptElement {
         }
 
         if tree_in_doc && !self.parser_inserted.get() {
-            self.prepare();
+            let script = Trusted::new(self);
+            document_from_node(self).add_delayed_task(task!(ScriptDelayedInitialize: move || {
```
[the remark this answers is not in the document]
Comment (nox) [reply components/script/dom/htmlscriptelement.rs]: This seems to specify the opposite.

https://html.spec.whatwg.org/multipage/scripting.html#script-processing-model:prepare-a-script-4

[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/htmlscriptelement.rs]: It still is immeditate, but it waits until the current DOM modification is complete, so no intermediate state can be observed by script or layout.

**Pull request merged**